packages/dapp-svelte-wallet/api/src/lib-wallet.js

```diff
@@ -482,70 +482,79 @@ export function makeWallet({
 
     const { inviteP, purseKeywordRecord, proposal } = await compiledOfferP;
 
-    // We now have everything we need to provide Zoe, so do the actual withdrawal.
-    // Payments are made for the keywords in proposal.give.
-    const keywords = [];
+    // Track from whence our the payment came.
+    /** @type {Map<Payment, Purse>} */
+    const paymentToPurse = new Map();
 
-    const paymentPs = Object.entries(proposal.give || harden({})).map(
-      ([keyword, amount]) => {
+    // We now have everything we need to provide Zoe, so do the actual withdrawals.
+    // Payments are made for the keywords in proposal.give.
+    const keywordPaymentPs = Object.entries(proposal.give || harden({})).map(
+      async ([keyword, amount]) => {
         const purse = purseKeywordRecord[keyword];
         assert(
           purse !== undefined,
           X`purse was not found for keyword ${q(keyword)}`,
         );
-        keywords.push(keyword);
-        return E(purse).withdraw(amount);
+        const payment = await E(purse).withdraw(amount);
+        paymentToPurse.set(payment, purse);
+        return [keyword, payment];
       },
     );
 
+    // Try reclaiming any of our payments that we successfully withdrew, but
+    // were left unclaimed.
+    const tryReclaimingWithdrawnPayments = () =>
+      // Use allSettled to ensure we attempt all the deposits, regardless of
+      // individual rejections.
+      Promise.allSettled(
+        keywordPaymentPs.map(async keywordPaymentP => {
+          // Wait for the withdrawal to complete.
+          const [_keyword, payment] = await keywordPaymentP;
+          // Find out where it came from.
+          const purse = paymentToPurse.get(payment);
+          // Now send it back to the purse.
+          // eslint-disable-next-line no-use-before-define
+          return addPayment(payment, purse);
+        }),
+      );
+
+    // Gather all of our payments, and if there's an error, reclaim the ones
+    // that were successfuly withdrawn.
+    const withdrawAllPayments = Promise.all(keywordPaymentPs);
+    withdrawAllPayments.catch(tryReclaimingWithdrawnPayments);
+
     // =====================
     // === AWAITING TURN ===
     // =====================
 
     // this await is purely to prevent "embarrassment" of
     // revealing to zoe that we had insufficient funds/assets
     // for the offer.
-    const payments = await Promise.all(paymentPs);
-
-    const paymentKeywordRecord = harden(
-      Object.fromEntries(keywords.map((keyword, i) => [keyword, payments[i]])),
-    );
+    const paymentKeywords = await withdrawAllPayments;
+    const paymentKeywordRecord = harden(Object.fromEntries(paymentKeywords));
 
     const seat = E(zoe).offer(inviteP, harden(proposal), paymentKeywordRecord);
```

Comment:
Zoe guarantees that the payments are successfully escrowed before the seat is returned, so you could do `tryReclaimingWithdrawnPayments();` right after this. That might be conceptually clearer to avoid confusing with payout handling.

Reply:
Since `seat` is a promise, you could do `seat.catch`

Reply:
> Zoe guarantees that the payments are successfully escrowed before the seat is returned, so you could do `tryReclaimingWithdrawnPayments();` right after this.

I'll make `tryReclaiming...` idempotent and add a `seat.finally` as well.  I still want to be maximally defensive, so I think the `depositedP.finally` should stand.  (Because of pipelining, it's possible the `getPayouts()` is successful even if `seat` never resolves.)